src/modules/recipe/containers/RecipeForm.tsx

```diff
@@ -0,0 +1,178 @@
+  const navigate = useNavigate();
+  const initialFormData = useInitialFormData(recipe);
+  const recipeId = recipe?.id;
+  const [formData, setFormData] = React.useState<Data>(initialFormData);
+
+  React.useEffect(() => {
+    setFormData(initialFormData);
+  }, [initialFormData]);
+
+  const handleSubmit = React.useCallback(
+    async (event: React.FormEvent<HTMLFormElement>) => {
+      event.preventDefault();
+      const recipe = sanitize(formData);
+
+      if (recipeId) {
+        await update(recipeId, recipe);
+        navigate(`/recipes/${recipeId}`);
+      } else {
+        await create(recipe);
+        navigate('/recipes');
+      }
+    },
+    [recipeId, formData],
+  );
+
+  const handleChange = React.useCallback<OnChange>((event) => {
+    setFormData((previousFormData) => ({
+      ...previousFormData,
+      [event.target.name]: event.target.value,
+    }));
+  }, []);
+
+  const fields = useRecipeFormFields({
+    onChange: handleChange,
+    values: formData,
+  });
+
+  return (
+    <RecipeForm
+      onSubmit={handleSubmit}
+      // @ts-ignore: OnChange event type not detected on textarea.
+      fields={fields}
+      SubmitProps={{ variant: 'success' }}
+    />
+  );
+};
+
+export default RecipeFormContainer;
+
+const useInitialFormData = (recipe?: Recipe) =>
+  React.useMemo(
+    () => ({
+      name: '',
+      description: '',
+      ingredients: '',
+      ...(recipe ? desanitize(recipe) : {}),
+    }),
+    [recipe],
+  );
+
+const useRecipeFormFields = ({
+  onChange,
+  values,
+}: {
+  onChange: OnChange;
+  values: Data;
+}) =>
+  React.useMemo(
+    () => ({
+      name: {
+        InputProps: {
+          id: 'name',
+          type: 'text',
+          name: 'name',
+          value: values['name'],
+          required: true,
+          onChange,
+        },
+        LabelProps: {
+          htmlFor: 'name',
+          children: 'Name',
+        },
+        HelperProps: {
+          children: 'E.g. Pizza Napoli',
+        },
+      },
+      description: {
+        TextAreaProps: {
+          id: 'description',
+          type: 'text',
+          name: 'description',
+          value: values['description'],
+          required: true,
+          onChange,
+        },
+        LabelProps: {
+          htmlFor: 'description',
+          children: 'Description',
+        },
+        HelperProps: {
+          children: 'Recipe description',
+        },
+      },
+      ingredients: {
+        TextAreaProps: {
+          id: 'ingredients',
+          type: 'text',
+          name: 'ingredients',
+          required: true,
+          value: values['ingredients'],
+          onChange,
+        },
+        LabelProps: {
+          htmlFor: 'ingredients',
+          children: 'Ingredients',
+        },
+        HelperProps: {
+          children: 'Recipe ingredients separated by space',
+        },
+      },
+    }),
+    [onChange, values],
+  );
+
+async function create(recipe: RecipeQuery) {
```

Comment:
those methods, which are the point of contact with our server, need to be in a separate module, here you're doing many thinks on the same file, and at different levels

Reply:
Got you